File: backend/rust/src/server/udp.rs
use std::net::SocketAddr;
use std::sync::atomic::Ordering::Relaxed;
use std::sync::Arc;
use std::time::{Instant, SystemTime, UNIX_EPOCH};

use tokio::net::UdpSocket;

use super::session::Session;
use super::Server;
use crate::protocol::udp::*;
use crate::routing;

/// Per-worker reusable buffers so the 20 ms hot path does not allocate.
pub(crate) struct Scratch {
    plain: Vec<u8>,
    relay_plain: Vec<u8>,
    outgoing: Vec<(SocketAddr, Vec<u8>)>,
    free: Vec<Vec<u8>>,
}

impl Scratch {
    pub fn new() -> Self {
        Scratch {
            plain: Vec::with_capacity(MAX_DATAGRAM),
            relay_plain: Vec::with_capacity(MAX_DATAGRAM),
            outgoing: Vec::new(),
            free: Vec::new(),
        }
    }

    fn buf(&mut self) -> Vec<u8> {
        let mut b = self.free.pop().unwrap_or_else(|| Vec::with_capacity(MAX_DATAGRAM + 64));
        b.clear();
        b
    }
}

pub(crate) async fn worker(srv: Arc<Server>, sock: Arc<UdpSocket>) {
    let mut buf = vec![0u8; MAX_DATAGRAM + 1];
    let mut scratch = Scratch::new();
    loop {
        let (n, from) = match sock.recv_from(&mut buf).await {
            Ok(v) => v,
            Err(_) => {
                if srv.shutdown.load(Relaxed) {
                    return;
                }
                continue;
            }
        };
        srv.metrics.packets_received.fetch_add(1, Relaxed);
        srv.metrics.bytes_received.fetch_add(n as u64, Relaxed);
        let start = Instant::now();
        let relayed = srv.handle_datagram(&buf[..n], from, &mut scratch);
        let mut outgoing = std::mem::take(&mut scratch.outgoing);
        for (addr, dg) in outgoing.drain(..) {
            if sock.send_to(&dg, addr).await.is_ok() {
                srv.metrics.packets_sent.fetch_add(1, Relaxed);
                srv.metrics.bytes_sent.fetch_add(dg.len() as u64, Relaxed);
            }
            scratch.free.push(dg);
        }
        scratch.outgoing = outgoing;
        if relayed {
            srv.metrics.observe_latency(start.elapsed().as_secs_f64());
        }
    }
}

impl Server {
    /// Process one datagram; queues responses into `scratch.outgoing`.
    /// Returns true if it was a voice frame that went through routing.
    pub(crate) fn handle_datagram(&self, dg: &[u8], from: SocketAddr, scratch: &mut Scratch) -> bool {
        let h = match parse_header(dg, DIR_C2S) {
            Ok(h) => h,
            Err(e) => {
                self.metrics.invalid(e.as_str());
                return false;
            }
        };
        let Some(sess) = self.hub.read().unwrap().by_conn.get(&h.connection_id).cloned() else {
            self.metrics.invalid("unknown_session");
            return false;
        };
        let now = Instant::now();
        let mut u = sess.udp.lock().unwrap();
        let is_cur = u.is_cur(h.key_id);
        let has_prev = u.prev.is_some();
        let Some(key) = u.key_for(h.key_id, now) else {
            self.metrics.invalid("unknown_key");
            return false;
        };
        if !key.window.check(h.counter) {
            self.metrics.drop("replay");
            return false;
        }
        if key.key.open(&mut scratch.plain, DIR_C2S, &h, dg).is_err() {
            self.metrics.invalid("auth_failed");
            return false;
        }
        if let Err(e) = parse_plaintext(h.ptype, &scratch.plain) {
            self.metrics.invalid(e.as_str());
            return false;
        }
        let new_top = key.window.update(h.counter);
        if is_cur && has_prev {
            u.send_with_new = true; // client proved it has the rotated key
        }
        match h.ptype {
            TYPE_HELLO => {
                let (uid, ct) = parse_hello(&scratch.plain).unwrap();
                if uid != sess.uuid_b {
                    self.metrics.invalid("uuid_mismatch");
                    return false;
                }
                let first = u.addr.is_none();
                u.addr = Some(from);
                let send_ok = !u.udp_ok_sent;
                u.udp_ok_sent = true;
                drop(u);
                if first {
                    self.metrics.active_voice_sessions.fetch_add(1, Relaxed);
                    if let Some(ps) = self.hub.write().unwrap().peers.get_mut(&sess.conn_id) {
                        ps.peer.udp_verified = true;
                    }
                }
                self.reply(&sess, TYPE_HELLO_ACK, ct, scratch);
                if send_ok {
                    sess.send(r#"{"type":"udp_ok"}"#.to_string());
                }
                false
            }
            TYPE_PING => {
                if u.addr.is_none() {
                    self.metrics.drop("no_hello");
                    return false;
                }
                if new_top {
                    u.addr = Some(from);
                }
                drop(u);
                let ct = parse_time(&scratch.plain).unwrap();
                self.reply(&sess, TYPE_PONG, ct, scratch);
                false
            }
            TYPE_VOICE => {
                if u.addr.is_none() {
                    self.metrics.drop("no_hello");
                    return false;
                }
                if new_top {
                    u.addr = Some(from); // NAT rebinding only on a fresh authenticated counter
                }
                if !u.voice_bucket.allow(now) {
                    self.metrics.drop("rate_limited");
                    return false;
                }
                let plain = std::mem::take(&mut scratch.plain);
                let v = parse_voice(&plain).unwrap();
                track_jitter(&mut u, v.timestamp, now);
                drop(u);
                self.relay_voice(&sess, &v, scratch);
                scratch.plain = plain;
                true
            }
            _ => false,
        }
    }

    fn reply(&self, sess: &Session, t: u8, ct: u64, scratch: &mut Scratch) {
        let mut out = scratch.buf();
        let mut u = sess.udp.lock().unwrap();
        let Some(addr) = u.addr else { return };
        let k = u.send_key(Instant::now());
        k.send_counter += 1;
        let h = Header {
            ptype: t,
            key_id: k.id,
            connection_id: sess.conn_id,
            counter: k.send_counter,
        };
        k.key.seal(&mut out, DIR_S2C, &h, &ct.to_be_bytes());
        scratch.outgoing.push((addr, out));
    }

    fn relay_voice(&self, sender: &Session, v: &Voice<'_>, scratch: &mut Scratch) {
        let now_ms = self.now_ms();
        let targets: Vec<(Arc<Session>, u32)> = {
            let hub = self.hub.read().unwrap();
            let Some(sp) = hub.peers.get(&sender.conn_id) else { return };
            if !routing::sender_eligible(&self.rcfg, now_ms, &sp.peer, v.epoch) {
                let reason = if sp.peer.in_world && v.epoch != sp.peer.epoch {
                    "stale_epoch"
                } else {
                    "not_routable"
                };
                self.metrics.drop(reason);
                return;
            }
            let Some(bucket) = hub.buckets.get(&sp.peer.scope_key) else {
                return;
            };
            bucket
                .iter()
                .filter_map(|c| {
                    let rp = hub.peers.get(c)?;
                    if routing::deliver(&self.rcfg, now_ms, &sp.peer, &rp.peer, v.mode) {
                        Some((hub.by_conn.get(c)?.clone(), rp.peer.epoch))
                    } else {
                        None
                    }
                })
                .collect()
        };
        let now = Instant::now();
        for (r, epoch) in targets {
            scratch.relay_plain.clear();
            write_relay(&mut scratch.relay_plain, &sender.uuid_b, epoch, v);
            let mut out = scratch.buf();
            let mut u = r.udp.lock().unwrap();
            let Some(addr) = u.addr else { continue };
            let k = u.send_key(now);
            k.send_counter += 1;
            let h = Header {
                ptype: TYPE_VOICE_RELAY,
                key_id: k.id,
                connection_id: r.conn_id,
                counter: k.send_counter,
            };
            k.key.seal(&mut out, DIR_S2C, &h, &scratch.relay_plain);
            drop(u);
            scratch.outgoing.push((addr, out));
        }
    }
}

/// RFC 3550 interarrival jitter estimate.
fn track_jitter(u: &mut super::session::UdpState, ts: u32, now: Instant) {
    let arrival = SystemTime::now().duration_since(UNIX_EPOCH).map(|d| d.as_secs_f64()).unwrap_or(0.0);
    let transit = arrival - ts as f64 / 48000.0;
    if let (Some(last), Some(lv)) = (u.last_transit, u.last_voice) {
        if now.duration_since(lv).as_secs_f64() < 1.0 {
            let d = (transit - last).abs();
            if d < 1.0 {
                u.jitter += (d - u.jitter) / 16.0;
            }
        }
    }
    u.last_transit = Some(transit);
    u.last_voice = Some(now);
}

File: backend/go/internal/server/udp.go
package server

import (
	"context"
	"encoding/binary"
	"errors"
	"net"
	"net/netip"
	"runtime"
	"sync"
	"time"

	"github.com/RavoxX/MCVoice/backend/go/internal/protocol"
	"github.com/RavoxX/MCVoice/backend/go/internal/routing"
)

var bufPool = sync.Pool{New: func() any { b := make([]byte, 0, 2048); return &b }}

// ListenUDP binds the voice socket.
func (s *Server) ListenUDP() error {
	addr, err := net.ResolveUDPAddr("udp", net.JoinHostPort(s.cfg.VoiceBindAddress, itoa(s.cfg.VoiceUDPPort)))
	if err != nil {
		return err
	}
	c, err := net.ListenUDP("udp", addr)
	if err != nil {
		return err
	}
	_ = c.SetReadBuffer(4 << 20)
	_ = c.SetWriteBuffer(4 << 20)
	s.udp = c
	return nil
}

// UDPAddr returns the bound voice address (useful for tests using port 0).
func (s *Server) UDPAddr() *net.UDPAddr { return s.udp.LocalAddr().(*net.UDPAddr) }

// ServeUDP runs the relay workers until ctx is cancelled.
func (s *Server) ServeUDP(ctx context.Context) {
	workers := s.cfg.UDPWorkers
	if workers <= 0 {
		workers = runtime.GOMAXPROCS(0)
	}
	var wg sync.WaitGroup
	for i := 0; i < workers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			s.udpWorker(ctx)
		}()
	}
	<-ctx.Done()
	_ = s.udp.Close()
	wg.Wait()
}

func (s *Server) udpWorker(ctx context.Context) {
	buf := make([]byte, protocol.MaxDatagram+1)
	pt := make([]byte, 0, protocol.MaxDatagram)
	for {
		n, from, err := s.udp.ReadFromUDPAddrPort(buf)
		if err != nil {
			if ctx.Err() != nil || errors.Is(err, net.ErrClosed) {
				return
			}
			continue
		}
		s.metrics.PacketsReceived.Inc()
		s.metrics.BytesReceived.Add(uint64(n))
		s.handleDatagram(buf[:n], from, pt[:0])
	}
}

func (s *Server) invalid(reason string) { s.metrics.InvalidPackets.With(reason).Inc() }
func (s *Server) dropped(reason string) { s.metrics.PacketsDropped.With(reason).Inc() }

func (s *Server) handleDatagram(dg []byte, from netip.AddrPort, pt []byte) {
	start := time.Now()
	h, err := protocol.ParseHeader(dg, protocol.DirC2S)
	if err != nil {
		s.invalid(protocol.ClassOf(err))
		return
	}
	s.mu.RLock()
	sess := s.byConn[h.ConnectionID]
	s.mu.RUnlock()
	if sess == nil {
		s.invalid("unknown_session")
		return
	}
	now := s.now()

	sess.umu.Lock()
	key := sess.keyFor(h.KeyID, now)
	if key == nil {
		sess.umu.Unlock()
		s.invalid("unknown_key")
		return
	}
	if !key.window.Check(h.Counter) {
		sess.umu.Unlock()
		s.dropped("replay")
		return
	}
	pt, err = key.aead.Open(pt, protocol.DirC2S, h, dg)
	if err != nil {
		sess.umu.Unlock()
		s.invalid("auth_failed")
		return
	}
	if err := protocol.ParsePlaintext(h.Type, pt); err != nil {
		sess.umu.Unlock()
		s.invalid(protocol.ClassOf(err))
		return
	}
	newTop := key.window.Update(h.Counter)
	if key == &sess.cur && !sess.sendWithNew && sess.prev.valid {
		sess.sendWithNew = true // client proved it has the rotated key
	}

	switch h.Type {
	case protocol.TypeHello:
		uid, ct, _ := protocol.ParseHello(pt)
		if uid != sess.uuidB {
			sess.umu.Unlock()
			s.invalid("uuid_mismatch")
			return
		}
		first := !sess.hasAddr
		sess.addr, sess.hasAddr = from, true
		sendOK := !sess.udpOKSent
		sess.udpOKSent = true
		sess.umu.Unlock()
		if first {
			s.metrics.ActiveVoiceSessions.Add(1)
			s.mu.Lock()
			sess.peer.UDPVerified = true
			s.mu.Unlock()
		}
		s.reply(sess, protocol.TypeHelloAck, ct)
		if sendOK {
			sess.send(marshal(protocol.Simple{Type: "udp_ok"}))
		}
		return
	case protocol.TypePing:
		if !sess.hasAddr {
			sess.umu.Unlock()
			s.dropped("no_hello")
			return
		}
		if newTop {
			sess.addr = from
		}
		ct := binary.BigEndian.Uint64(pt)
		sess.umu.Unlock()
		s.reply(sess, protocol.TypePong, ct)
		return
	case protocol.TypeVoice:
		if !sess.hasAddr {
			sess.umu.Unlock()
			s.dropped("no_hello")
			return
		}
		if newTop {
			sess.addr = from // NAT rebinding, only on a fresh authenticated counter
		}
		if !sess.voiceBucket.allow(now) {
			sess.umu.Unlock()
			s.dropped("rate_limited")
			return
		}
		v, _ := protocol.ParseVoice(pt)
		s.trackJitter(sess, v.Timestamp, now)
		sess.umu.Unlock()
		s.relayVoice(sess, v, now)
		s.metrics.RelayLatency.Observe(time.Since(start).Seconds())
	}
}

// trackJitter maintains an RFC 3550 interarrival jitter estimate (caller holds umu).
func (s *Server) trackJitter(sess *Session, ts uint32, now time.Time) {
	arrival := float64(now.UnixNano()) / 1e9
	transit := arrival - float64(ts)/48000.0
	if sess.haveTransit && now.Sub(sess.lastVoice) < time.Second {
		d := transit - sess.lastTransit
		if d < 0 {
			d = -d
		}
		if d < 1 { // ignore timestamp wrap / talk-spurt restarts
			sess.jitter += (d - sess.jitter) / 16
		}
	}
	sess.lastTransit, sess.haveTransit, sess.lastVoice = transit, true, now
}

func (s *Server) reply(sess *Session, t byte, ct uint64) {
	var p [8]byte
	binary.BigEndian.PutUint64(p[:], ct)
	bp := bufPool.Get().(*[]byte)
	sess.umu.Lock()
	k := sess.sendKey(s.now())
	k.sendCounter++
	out := k.aead.Seal((*bp)[:0], protocol.DirS2C, protocol.Header{Type: t, KeyID: k.id, ConnectionID: sess.connID, Counter: k.sendCounter}, p[:])
	addr := sess.addr
	sess.umu.Unlock()
	s.write(out, addr)
	*bp = out
	bufPool.Put(bp)
}

func (s *Server) write(b []byte, addr netip.AddrPort) {
	if _, err := s.udp.WriteToUDPAddrPort(b, addr); err == nil {
		s.metrics.PacketsSent.Inc()
		s.metrics.BytesSent.Add(uint64(len(b)))
	}
}

type target struct {
	sess  *Session
	epoch uint32
	group bool // relayed with mode 2 (spec 8.1)
}

var targetPool = sync.Pool{New: func() any { t := make([]target, 0, 16); return &t }}

func (s *Server) relayVoice(sender *Session, v protocol.Voice, now time.Time) {
	nowMs := now.UnixMilli()
	tp := targetPool.Get().(*[]target)
	targets := (*tp)[:0]

	s.mu.RLock()
	group := routing.GroupApplies(&sender.peer, v.Mode, v.Flags)
	proximity := v.Mode != protocol.ModeGroup && routing.SenderEligible(&s.rcfg, nowMs, &sender.peer, v.Epoch)
	if !group && !proximity {
		reason := "not_routable"
		if sender.peer.InWorld && v.Epoch != sender.peer.Epoch {
			reason = "stale_epoch"
		}
		s.mu.RUnlock()
		s.dropped(reason)
		*tp = targets
		targetPool.Put(tp)
		return
	}
	if group {
		// spec 8.1: every other member of the sender's group, relayed with mode 2
		if g := s.groups[sender.peer.Group]; g != nil {
			for _, r := range g.members {
				if routing.DeliverGroup(&sender.peer, &r.peer) {
					targets = append(targets, target{r, r.peer.Epoch, true})
				}
			}
		}
	}
	if proximity {
		// candidates are the players the sender's own world tracks (mutual visibility is mandatory)
		for u := range sender.peer.Visible {
			if r := s.byUUID[u]; r != nil && routing.DeliverProximity(&s.rcfg, nowMs, &sender.peer, &r.peer, v.Mode, group) {
				targets = append(targets, target{r, r.peer.Epoch, false})
			}
		}
	}
	s.mu.RUnlock()

	if len(targets) > 0 {
		bp := bufPool.Get().(*[]byte)
		ptb := bufPool.Get().(*[]byte)
		for _, t := range targets {
			rv := v
			if t.group {
				rv.Mode, rv.Flags = protocol.ModeGroup, v.Flags&^protocol.FlagGroup
			}
			plain := protocol.AppendRelay((*ptb)[:0], sender.uuidB, t.epoch, rv)
			t.sess.umu.Lock()
			if !t.sess.hasAddr {
				t.sess.umu.Unlock()
				continue
			}
			k := t.sess.sendKey(now)
			k.sendCounter++
			out := k.aead.Seal((*bp)[:0], protocol.DirS2C,
				protocol.Header{Type: protocol.TypeVoiceRelay, KeyID: k.id, ConnectionID: t.sess.connID, Counter: k.sendCounter}, plain)
			addr := t.sess.addr
			t.sess.umu.Unlock()
			s.write(out, addr)
			*bp, *ptb = out, plain
		}
		bufPool.Put(bp)
		bufPool.Put(ptb)
	}
	for i := range targets {
		targets[i] = target{}
	}
	*tp = targets[:0]
	targetPool.Put(tp)
}

func itoa(n int) string {
	var b [20]byte
	i := len(b)
	if n == 0 {
		return "0"
	}
	neg := n < 0
	if neg {
		n = -n
	}
	for n > 0 {
		i--
		b[i] = byte('0' + n%10)
		n /= 10
	}
	if neg {
		i--
		b[i] = '-'
	}
	return string(b[i:])
}
